src/cmdstan/command.hpp

```diff
@@ -187,9 +193,61 @@ int command(int argc, const char *argv[]) {
       diagnostic_csv_writers;
   std::vector<stan::callbacks::json_writer<std::ofstream>>
       diagnostic_json_writers;
+  std::vector<stan::callbacks::json_writer<std::ofstream>> metric_json_writers;
+
+  std::string output_base = get_basename_suffix(output_file).first;
+  std::string diagnostic_base;
+  if (!diagnostic_file.empty())
+    diagnostic_base = get_basename_suffix(diagnostic_file).first;
+  else
+    diagnostic_base = output_base;
+  bool save_single_paths
+      = user_method->arg("pathfinder")
+        && (get_arg_val<bool_argument>(parser, "method", "pathfinder",
+                                       "save_single_paths"));
+  bool save_diagnostics = diagnostic_base != output_base;
 
-  init_callbacks(parser, sample_writers, diagnostic_csv_writers,
-                 diagnostic_json_writers);
+  if (user_method->arg("pathfinder")) {
+    if (num_chains == 1) {
+      init_filestream_writers(sample_writers, num_chains, id, output_base, "",
+                              ".csv", sig_figs, "# ");
+      if (save_single_paths || save_diagnostics) {
+        init_filestream_writers(diagnostic_json_writers, num_chains, id,
+                                diagnostic_base, "", ".json", sig_figs);
+      } else {
+        init_null_callbacks(diagnostic_json_writers, num_chains);
+      }
+    } else {
+      if (save_single_paths || save_diagnostics) {
+        init_filestream_writers(sample_writers, num_chains, id, output_base,
+                                "_path", ".csv", sig_figs, "# ");
+        init_filestream_writers(diagnostic_json_writers, num_chains, id,
+                                diagnostic_base, "_path", ".json", sig_figs);
+      } else {
+        init_null_callbacks(sample_writers, num_chains);
+        init_null_callbacks(diagnostic_json_writers, num_chains);
+      }
+    }
+    init_null_callbacks(diagnostic_csv_writers, num_chains);
+  } else {
+    init_filestream_writers(sample_writers, num_chains, id, output_file, "",
+                            ".csv", sig_figs, "# ");
+    if (save_diagnostics) {
+      init_filestream_writers(diagnostic_csv_writers, num_chains, id,
+                              diagnostic_base, "", ".csv", sig_figs, "# ");
+    } else {
+      init_null_callbacks(diagnostic_csv_writers, num_chains);
```

Comment:
[Comment] If we ever get uncomfortable with the amount of nesting in this config, we could write a `conditional_init_writers` function which takes in a bool and all the remaining arguments and does this common if-else pattern.

Reply:
so far, just 2.   it's a toss up.